exercises/practice/two-bucket/.meta/example.ex

```diff
@@ -0,0 +1,55 @@
+defmodule TwoBucket do
+  defstruct [:goal, :moves, :other_bucket]
+  @type t :: %TwoBucket{goal: :one | :two, moves: integer, other_bucket: integer}
+
+  @doc """
+  Find the quickest way to fill a bucket with some amount of water from two buckets of specific sizes.
+  """
+  @spec measure(
+          size_one :: integer,
+          size_two :: integer,
+          goal :: integer,
+          start_bucket :: :one | :two
+        ) :: {:ok, TwoBucket.t()} | {:error, :impossible}
+  def measure(size_one, size_two, goal, start_bucket) do
+    other_filled = if start_bucket == :one, do: {0, size_two}, else: {size_one, 0}
+    forbidden_states = MapSet.new([other_filled])
+
+    # Partially apply function to avoid passing many arguments
+    next_moves = &next_moves(&1, size_one, size_two)
+
+    pour([{{0, 0}, 0}], forbidden_states, next_moves, goal)
+  end
+
+  def pour([], _, _, _), do: {:error, :impossible}
+
+  def pour([{{goal, other}, moves} | _], _, _, goal),
+    do: {:ok, %TwoBucket{goal: :one, moves: moves, other_bucket: other}}
+
+  def pour([{{other, goal}, moves} | _], _, _, goal),
+    do: {:ok, %TwoBucket{goal: :two, moves: moves, other_bucket: other}}
+
+  def pour([{state, moves} | states], forbidden_states, next_moves, goal) do
+    next =
+      next_moves.(state)
+      |> Enum.reject(&MapSet.member?(forbidden_states, &1))
+      |> Enum.map(&{&1, moves + 1})
+
+    pour(states ++ next, MapSet.put(forbidden_states, state), next_moves, goal)
+  end
+
+  def next_moves({fill_one, fill_two}, size_one, size_two) do
+    [
+      # Empty a bucket
+      {0, fill_two},
+      {fill_one, 0},
+      # Fill a bucket
+      {fill_one, size_two},
+      {size_one, fill_two},
+      # Pour one into the other
+      {min(size_one, fill_one + fill_two), max(0, fill_two - size_one + fill_one)},
+      {max(0, fill_one - size_two + fill_two), min(size_two, fill_one + fill_two)}
```

Comment:
Normally I don't comment on the example solution, but I'm actually struggling to understand this exercise. Could you explain to me why `min` and `max` are necessary? I would think that when you pour from one bucket into the other until the first one is full, there is only one possible outcome.

Reply:
No worries, I'm happy to explain. There is only one possible outcome indeed, but you have to be careful how to calculate it  and I'm using min/max as a shortcut not to use an `if` statement. For example, imagine we want to pour the first bucket in the other one (line 51). There are 2 cases:
1) There is enough room in the second bucket for all the water => result is 0 in the first bucket
2) There is not enough room in the second bucket => result is current fill `fill_one` minus the remaining room in the second bucket `size_two - fill two`. 
In general, I can calculate `fill_one - (size_two - fill_two)` and if it's negative, it means that there was extra room in the second bucket but since columns can't be negative, I need to cap that to 0. Hence `max(0, fill_one - (size_two - fill_two))`. If the same as `if fill_one > size_two - fill_two, do: 0, else: fill_one - (size_two - fill_two) `.

It's a similar process for the other bucket, but I use min to prevent over-fullness. 
Does that make sense? Should I use an `if` statement instead?

Reply:
Makes sense now, thanks! No, there's no need to change it to an `if` :)